src/work/ConditionalWork.cpp

```diff
@@ -0,0 +1,100 @@
+// Copyright 2019 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "ConditionalWork.h"
+#include "util/Logging.h"
+#include "util/format.h"
+
+namespace stellar
+{
+
+ConditionalWork::ConditionalWork(Application& app, std::string name,
+                                 ConditionFn condition,
+                                 std::shared_ptr<BasicWork> conditionedWork,
+                                 std::chrono::milliseconds sleepTime)
+    : BasicWork(app, std::move(name), BasicWork::RETRY_NEVER)
+    , mCondition(std::move(condition))
+    , mConditionedWork(std::move(conditionedWork))
+    , mSleepDelay(sleepTime)
+    , mSleepTimer(std::make_unique<VirtualTimer>(app.getClock()))
+{
+    if (!mConditionedWork)
+    {
+        throw std::invalid_argument("Must provide valid work");
+    }
+
+    if (!mCondition)
+    {
+        throw std::invalid_argument("Must provide valid condition function. "
+                                    "Use BasicWork if condition is not "
+                                    "needed.");
+    }
+}
+
+BasicWork::State
+ConditionalWork::onRun()
+{
+    if (mWorkStarted)
+    {
+        mConditionedWork->crankWork();
+        return mConditionedWork->getState();
```

Comment:
I don't think this is correct:
`ConditionalWork`'s state will stay as `RUNNING` (even if `mConditionedWork` transitions to other states like WAITING).
Easiest might be to have conditional work derive from `Work` to deal with all the state transitions and proper sleeping.

Reply:
Clarified offline: `ConditionalWork` follows the state of `mConditionedWork`, so if the latter one transitions into a new state, the former work will just return that.  So this codepath should be right (we already do something similar in WorkSequence also)